[In Progress] feature: Porting mcuboot to Cortex-M0 with no VTOR support

miyatsu opened the pull request:
When I port mcuboot into zephyr on nrf51822 chip, I find the 51822 did not support vector table map register like stm32f0.

After decompiled the "softdevice S110" (Official Bluetooth Firmware provide by Nordic), I find the vector table is forward to bootloader or application by some sort of forward function.

Here is some test code and comment I wrote, please refer to the README.md file at https://github.com/miyatsu/mcuboot/tree/v0.9-branch-dev
Here is the code I tested successful. https://github.com/miyatsu/mcuboot/commit/6773e904c4aeb82140852588808c3dbdc2e85c29
By reading zephyr's source code, I think I may have the ability to solve this bug.

Before I doing this, I really need you guys to check my code for sure if this can doing well.
Thinks...

Note: All changes on this pull request are NOT tested, DO NOT MERGE!!!

Signed-off-by: Ding Tao <miyatsu@qq.com>

cvinayak commented:
Hi miyatsu,

In the SoftDevice, a binary component placed at the flash start performs the vector table forwarding either to SoftDevice or application or bootloader binary. This component is hardly ever reflashed under DFU.

In the case of Zephyr, mcuboot should be responsible for vector table forwarding to a Zephyr binary. Rationale, we do not want to reflash the vector table forwarding as part of flashing/downloading new Zephyr binary.

That said, I am not a mcuboot expert.

miyatsu commented:
Hi cvinayak,

That's what I'm doing here: write a fixed vector table and interrupt forward function placed at the very beginning of flash address 0x0. This is just like SoftDevice do, except of MSP(0x00000000) and RESET(0x00000004).

The interrupt forward function will forward every single irq to a copy of vector table in RAM, this vector table in RAM will be overwrite in function relocate_vector_table() after zephyr power up. Could be the mcuboot or pure zephyr kernel.

The table and the forward function will be linked together with the mcuboot. There is no need reflash this fixed binary component.

cvinayak commented:
@miyatsu 

The forwarder has to be in the bootloader (mcuboot or Zephyr bootloader or something else) and not in Zephyr Samples/Applications unless your intention is to use Zephyr as bootloader CONFIG_IS_BOOTLOADER=y (or secondary program loader) to another Zephyr Sample/Application.

One, your implementation to use Zephyr as Primary Program Loader for Cortex M0 with no VTOR is welcome, but its not restricted to nRF51, your PR will need to be towards a generic solution.

Two, the forwarder shall not be restricted to the vector table being in RAM, the target binary will have the vector table in flash, in most cases a copy to RAM is unnecessary.

Three, does your branch change the flash map in mcuboot? https://github.com/runtimeco/mcuboot/blob/master/docs/design.md#flash-map
I see in your branch you have added a forwarder to mcuboot, these should suffice it in loading a Zephyr image built for flash area image 0.

nvlsianpu commented:
1st) I would hange the request name - this patch is adding functionality, not fixing any bug. I sugest something like: emulation of vector relocation for thargets whitch don't support it in HW.

2nd) I would use value of ```CONFIG_FLASH_LOAD_OFFSET + CONFIG_TEXT_SECTION_OFFSET``` (1st comes from DTS, 2nd form Kconfig) here for enable/disable software vector relocation for targets whitch don't support vectror relocation by HW.

3rd) I will put more coments soon...

nvlsianpu commented:
@miyatsu @cvinayak 
I'm appreciate you propose this patch.

For clarification, mcuboot starts the image application by directly call to its reset vector. Call address is known by ```CONFIG_FLASH_LOAD_OFFSET + CONFIG_TEXT_SECTION_OFFSET``` briefly.

For those targets which have HW vector relocation the _reset_vecttor call of the Application, does the work (does the vector address set).

For cortex M0 vector relocation must be entirely done by mcuboot. This should be done by a redirecting function which would uses vector address (link) to get the beginning of the vector array  (which is placed in application image) thanks to knowledge of the application's _reset_vector  address (look https://github.com/runtimeco/mcuboot/blob/master/boot/zephyr/main.c#L57).

Proposed redirected function should initially redirecta IRQ handlers to vector of mcuboot, then just before calling the app-start should be reconfigured to redirect to the vector of application.

Summary:
In zephyr's mcuboot need a run-time vectror redirect function. Mcuboot's vector must be shifted and used by the redirect function initially. This should be incorporated into zephyr patch with appropriate Kconfig.
In zephyr' application only vector array need to be offset as it is already doable for cortex M4. For M0 there is no sense of call to relocate_vector_table(void) by the application.

@miyatsu Can you consider my proposition and rework this path.

cvinayak commented:
@nvlsianpu I agree with your proposal (as in your summary).

miyatsu commented:
@cvinayak @nvlsianpu 
1). I agree with your idea. This patch should towards to generic solution for those SoC without HW support for vector table relocation.

2). The run-time redirection without copy the vector table into RAM... I thought about it before I'm write those code, by reading the source code of "nrf51822_v4.4 - Bootloader_Without_SD.zip"(This can be find here: https://devzone.nordicsemi.com/question/33621/how-to-make-the-nrf51822-flash-jump-successful-without-softdevice/?answer=33678#post-id-33678 ).
It is done by checking a value in RAM, which has a fixed address, then decide which vector table should the bootloader jump (Bootloader it self, or Application?).
I just don't like the way it implemented personally.

3). Is that possible, make the IRQ redirection all done by mcuboot itself without access the RAM ?
I don't think so. Then what makes it so different to copy the whole vector table into RAM, instead of check one flag in RAM ?. That flag in RAM may be access(write) by application some how, CRASH will happen. (Although the whole vector table in RAM may also be changed by accident, I did not realize that at that time......Hahaha....)

4). I just graduated from college this year with limited knowledge. And I'm new to this project . I'm afraid I can't done well with the whole things you guys suggested in a short time. Three to six month or a year may needed by take all my weekend ?.. I'm not sure.

5).More comment will be added in next week, good night (goot afternoon in your timezone? hahaha).

b0661 commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -181,9 +190,10 @@ SECTIONS
 
-#if defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_IS_BOOTLOADER)
+#if defined (CONFIG_SOC_SERIES_NRF51X) ||	\
+	defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_IS_BOOTLOADER)
```
Parenthesis preferred to make grouping clear (and avoid mistakes).

    #if (defined(CONFIG_SOC_SERIES_NRF51X) || \
          defined(CONFIG_SOC_SERIES_STM32F0X)) \
         && !defined(CONFIG_IS_BOOTLOADER)

b0661 commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -181,9 +190,10 @@ SECTIONS
+	defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_IS_BOOTLOADER)
     /* Must be first in ramable region */
-	SECTION_PROLOGUE(.st_stm32f0x_vt,(NOLOAD),)
+	SECTION_PROLOGUE(.fixed_vector_table,(NOLOAD),)
```
The name .fixed_vector_table maybe misleading. It is not fixed in the sense of const. Maybe call it .ram_vector_table.

nvlsianpu commented:
@miyatsu 
2) It can bee achieved using more sophisticated method. Function which redirects IRQ handler need only to know the IRQ number and where vectors table is.

>3). Is that possible, make the IRQ redirection all done by mcuboot itself without access the RAM ?
I don't think so. Then what makes it so different to copy the whole vector table into RAM, instead of check one flag in RAM ?. That flag in RAM may be access(write) by application some how, CRASH will happen. (Although the whole vector table in RAM may also be changed by accident, I did not realize that at that time......Hahaha....)

You are right: Mcuboot need 4B of RAM for memorize address for pointer to vectors table. It is still much less than whole table copy in RAM - vectors table can be placed in RAM too if someone decide to for a reason - but this is still independent option.

4) I will take it into account :D, 👍 
5) My timezone is CET.

carlescufi commented:
First, thanks for your work on this.
I think we should start with a new configuration option that defines whether the core supports a relocatable vector table in hardware or not, something along the lines of:
`CONFIG_CPU_HAS_RELOC_VT`
That way we can activate it for M3 and M4 but leave it to `n` for M0, and use this as switch mechanism to know where we need to relocate, in the following proposed manner for Cortex-M0 in mcuboot:
```
    /* Lock interrupts and dive into the entry point */
    irq_lock();
#if !defined(CONFIG_CPU_HAS_RELOC_VT)
    /* Set an internal variable that turns interrupt forwarding to the Application image */
    relocate_vector_table();
#endif
    ((void (*)(void))start)();
```

superna9999 commented:
On STM32, we simply had to apply the fix in #5427 to allow using applications with bootloader.
The NFR relocatable vector should be the same as STM32F0/L0, no need for an IRQ forwarder.

miyatsu commented:
Hi, @superna9999 
Thanks for the comment, I don't think your PR#5427 would working on NRF51, check this article: https://devzone.nordicsemi.com/question/139887/interrupt-vectors-remap/

miyatsu commented:
I've refactor the code, but not tested yet (No hardware available right now), please review. @carlescufi @nvlsianpu @cvinayak

miyatsu commented on arch/arm/core/cortex_m/vector_table.S:
```diff
@@ -69,3 +69,80 @@ SECTION_SUBSEC_FUNC(exc_vector_table,_vector_table_section,_vector_table)
     .word __reserved
 #endif
+
+#if !defined(CONFIG_CPU_CORTEX_M_HAS_VTOR)
```
I think here need change to
#if ( !defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && defined(CONFIG_IS_BOOTLOADER) )

miyatsu commented on arch/arm/core/cortex_m/vector_table.S:
```diff
@@ -69,3 +69,80 @@ SECTION_SUBSEC_FUNC(exc_vector_table,_vector_table_section,_vector_table)
+GDATA(_ram_vector_start)
+
+SECTION_FUNC(vector_relay_handler, __vector_relay_handler)
+    mrs     r0, ipsr;       /* Load interript program status register into r0 */
```
Is there need to lock the irq by "cpsid i" before run "mrs r0, ipsr" ?

nvlsianpu commented:
cc @ioannisg

superna9999 commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -35,6 +35,14 @@ config ISA_THUMB2
 	  technology is featured in the  processor, and in all ARMv7
 	  architecture-based processors.
 
+config CPU_CORTEX_M_HAS_VTOR
```
HAS_VTOR is not right, STM32 does not has VTOR, it can remap the 0x0 physical address space with either the flash or the internal RAM.
A more coherent name would be CPU_CORTEX_M_CAN_RELOCATE for M0

miyatsu replied:
How about "HAS_FIXED_VT_ADDR" ?

superna9999 replied:
Yes for M0 only then

nvlsianpu replied:
I think CPU_CORTEX_M_HAS_VTOR is proper for native cortex vector relocation support - it also well References to the cortex manual. Looks like STMF0 has much different mechanism for vector relocation than VITOR is.

So maybe ther should be 2 configuration Keyword, on for informing that there is support relocation feature support, and 2nd for describe mechanism used (like VTOR is).

miyatsu replied:
I've checked most of the M0 SoCs spec datasheet (NXP L11XX/12XX etc.), it seems only STM32F0 support such a vector table remap mechanism.

All M0 SoC are all have fixed vector table address at 0x00000000, "HAS_FIXED_VT_ADDR" may not goot for this.

I think we use only one configuration is enough for this, enable it only in M0 SoCs but STM32F0.

Vector table relocation only enabled when SoC support bootlaoder, we can write like this:

In arch/arm/soc/{M0 SoCs}/Kconfig file:
select CPU_CORTEX_M0_VT_RELAY if (IS_BOOTLOADER || IMG_MANAGER)

As for STM32F0 Kconfig:
select CPU_CORTEX_M0_VT_REMAP if (IS_BOOTLOADER || IMG_MANAGER)
/* 2nd configuration, unnecessary, we can just leave it blank */

The issue https://github.com/zephyrproject-rtos/zephyr/issues/5651 can write linker like this:
linker script:
    ...
#ifdef CPU_CORTEX_M0
    . = 0x0;

#ifdef CPU_CORTEX_M0_VT_RELAY 
    KEEP(Relay VT)
    KEEP(Relay handler)
#else /* Some HW mechanism support vector table relocation like remap in STM32F0*/
    KEEP(Normal VT)
#endif 

    . = . + CONFIG_TEXT_SECTION_OFFSET
#else /* None M0 SoC, vector table can place any where */
    . = CONFIG_TEXT_SECTION_OFFSET;
    KEEP(Normal VT)
#endif 
    ....

carlescufi commented:
@ioannisg, @skordal FYI

miyatsu committed:
arch: arm: Add CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP Kconfig flag

Cortex-M0 do not have the Vector Table Base Address Offset Register, so
Cortex-M0 vector table address can not be changed.

But in some Cortex-M0 SoCs like STM32F0 series, they have some mechanism
that can remap the vector table address to the start address of SRAM.

Use this flag to indicates whether current Cortex-M0 SoC support such a
remap or not.

Signed-off-by: Ding Tao <miyatsu@qq.com>

miyatsu committed:
arch: arm: stm32f0: Enable CONFIG_CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP

Enable this flag for all stm32f0 series in order to support hardware
vector table relocation feature.

Signed-off-by: Ding Tao <miyatsu@qq.com>

miyatsu commented:
@superna9999 @nvlsianpu Check the new patch.

miyatsu commented:
Hi, @carlescufi @nvlsianpu @superna9999 , sorry for bothering you:

I have some questions for the auto checks, it says I have ```uc6 body has no content, should at least have 1 line``` error at commit id ```ab3e6f5474```, but I can't find this commit id in github or my local repo.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf51/soc.c:
```diff
@@ -34,6 +36,17 @@ static bool ftpan_59(void);
 
 uint32_t SystemCoreClock __used = __SYSTEM_CLOCK;
 
+void relocate_vector_table(void)
```
Shouldn't this function be in a more generic file shared by all Cortex-M0 boards that do not have hw vector table relocation?

carlescufi replied:
I assume that this is the point where forwarding actually begins, from the bootloader to the application image, so it really makes sense that we have this somewhere in `arch/arm/cortex_m/`

carlescufi replied:
Actually, why add this if there's an identical version already present here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arm/core/cortex_m/prep_c.c#L30

miyatsu replied:
Good idea, will do.

carlescufi commented:
@andrewboie you might want to take a look at this one since it touches upon the IRQ management

nvlsianpu commented on arch/arm/soc/nordic_nrf5/nrf51/soc.c:
```diff
@@ -34,6 +36,17 @@ static bool ftpan_59(void);
 
+void relocate_vector_table(void)
+{
+#if defined(CONFIG_IS_BOOTLOADER) || defined(CONFIG_IMG_MANAGER)
```
There's no reason to connect vector relocation feature with CONFIG_IMG_MANAGER - it is different functionalities. image management itself won't use vector relocation. vector relocation call is needed for boot loader only.

nvlsianpu replied:
Except if you expect that vector relocation call should be call by runing application - then question is how application will know where placed bootloader's _ram_vector_start[] is?

miyatsu replied:
Well, I find zephyr has three way to boot.
And I design it like this:
1).Boot as a bootloader
Boot as bootloader, zephyr must have a vector relay function put at address 0x00000000, we use 0x20000000 as proxy to relay irq. And application and bootloader it self must copy its own vector table to 0x20000000 when boot up.

2).Chainload by bootloader
If chainload by bootloader, the application must copy its vector table to SRAM at 0x20000000 so the bootloader's irq relay handler can find.

3).Boot directly.
As for boot directly, the vector table must put at address 0x0, no need to copy the vector table to SRAM since there is no bootloader (aka vector relay handler) to use.

nvlsianpu replied:
reply to 1) bootloader can have the relay function implemented in such a way that it only needs the pointer (call it <b>vr_pointer</b>) to current vector table. So bootloader will have one HW supported vector table at adress 0x00, and the 2nd one somewhere else (for instance just after the HW one). Initiali the pointer will point at the 2nd vector table.

reply to 2)
Application should have its vector table where it use to be placet normally. So it will be possibile to just update <b>vr_pointer</b> for relocate the vector. It is possible that this will be done by a) bootloader -just before it calls reset vector of application. b) in application reset vector.

What I described will decrease consumption of RAM for VR - usually M0 targets have not much of it.

miyatsu replied:
No,,, I don't want to implemented like that. Its, ugly I think ? I really don't like that way, use one flag to identify current irq is for bootloader or for application. Too bad.. too bad..

Only 192 byte, not too much. One single ISR will need at lease 600+ bytes for its stack.

nrf51822 QFAA have 16KB SRAM, the vector table in only occupied 1.1 percent of total SRAM.

miyatsu replied:
Let me guess, you may say: If you don't change it like what I said, your PR will never be merged into master branch. Hahaha

I will try it, I will @you when it's done. @nvlsianpu

pfalcon replied:
@miyatsu :

> If you don't change it like what I said, your PR will never be merged into master branch.

Let me chime in, as a mere developer, just you guys. I don't think that @nvlsianpu could say that, at least he doesn't have merge rights, just the same as you or me don't have. He can only either approve your PR, not approve, or -1. Generally, -1'ed PR wouldn't be merged by the maintainers with write access. You definitely could try to instead get approvals from other folks and then escalate merging even despite somebody's -1. But much more productive way is to listen to suggestions of other developers, and try to understand their point of view. Over time, as you get more experience with Zephyr, you may see that the suggestions were indeed useful/right.

Definitely please don't feel that you're "cornered" into making some change. The idea is to cooperate to write the *better* code than a lone developer could do themselves.

P.S. You're doing great with this PR and review, thanks for your contribution and patience, much appreciated!

miyatsu replied:
Oh, I'm just kidding. Sorry for that...

nvlsianpu replied:
@miyatsu - how cooperation looks like was described very good by @pfalcon.
You decide as contributor what you will develop, but indeed you need to get approval of your changes - not only from me. I was getting annoyed by heavy and long review many times. Functionality you try to add is extremely important to us (Nordic semiconductor I mean) and any cortex-M0 vendor - that why I em so demanding. (So Many times we were fighting for a few bytes of RAM more for the application)
BTW - I meant nota a switch, but the pointer.
Flexible vector relocation like this provided by CMSI void NVIC_SetVector(IRQn_Type IRQn, uint32_t vector) is another feature than expected here.

miyatsu commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -184,9 +234,9 @@ SECTIONS
-	SECTION_PROLOGUE(.st_stm32f0x_vt,(NOLOAD),)
+	SECTION_PROLOGUE(.ram_vector_table,(NOLOAD),)
 	{
         _ram_vector_start = .;
```
Here @nvlsianpu , application and bootloader both need to reserve this memory array. When application is chainloaded, the prep_C() will put the vector table here.

andrewboie commented:
I got asked to review this -- unfortunately my knowledge of ARM is not extensive and although I don't see anything too crazy here (other that what has already been noted by other reviewers) I would defer to my Linaro/NXP colleagues, who I have added to the review request.

carlescufi commented:
See also related comments here: https://github.com/runtimeco/mcuboot/pull/166

carlescufi commented:
@miyatsu 

Could you clarify these points:
1. I don't see how the bootloader image's relay vector table knows about the application image's vector table address in flash
2. I don't see how the bootloader image's relay vector table is correctly switching from forwarding the exceptions and interrupts to the bootloader image's vector table to the application image's vector table.

Taken from here:
https://github.com/zephyrproject-rtos/zephyr/issues/5749 and comment here regarding the questions

miyatsu commented:
Wait,, I want to try the way @nvlsianpu suggested, until it’s tested perfectly.

carlescufi commented:
@miyatsu I think we need a new Kconfig variable, something like `CONFIG_CPU_CORTEX_M0_FORCE_RAM_VECTOR_TABLE`. @nvlsianpu what do you think?

nvlsianpu commented:
good keyword xx_FORCE_xx or xx_USING_x are both OK.

carlescufi commented:
Right, and then you could even enable `CONFIG_CPU_CORTEX_M0_FORCE_RAM_VECTOR_TABLE` it by default if building with `CONFIG_IMG_MGR` or with `CONFIG_TEXT_SECTION_OFFSET > 0`, to help out most people using mcuboot and Zephyr.

carlescufi commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -26,13 +26,19 @@
 #ifdef CONFIG_ARMV6_M
 
+#if defined(CONFIG_IRQ_SOFT_RELAY)
+_GENERIC_SECTION(.vt_pointeri_section) uint32_t *_vector_table_pointer;
```
`.vt_pointer_section`, remove the extra "i"

carlescufi commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -194,6 +203,15 @@ SECTIONS
+	SECTION_PROLOGUE(.vt_pointer,(NOLOAD),)
+	{
+		*(.vt_pointer_section)
+		*(".vt_pointer_sectioin.*")
```
`vt_pointer_section` remove the extra "i"

carlescufi commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -26,13 +26,19 @@
 	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
 	memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
+#elif defined(CONFIG_IRQ_SOFT_RELAY)
+	*_vector_table_pointer = (uint32_t)_vector_start;
```
This should be: 

`_vector_table_pointer = (void*) _vector_start`

carlescufi commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -26,13 +26,19 @@
 #ifdef CONFIG_ARMV6_M
 
+#if defined(CONFIG_IRQ_SOFT_RELAY)
+_GENERIC_SECTION(.vt_pointer_section) uint32_t *_vector_table_pointer;
```
This should be `void *_vector_table_pointer;`

carlescufi commented on arch/arm/core/irq_relay.S:
```diff
@@ -0,0 +1,112 @@
+	ldr		r1, = _vector_table_pointer;
+	ldr		r1, [r1];
+	adds	r1, r1, r0;
+
```
You are missing another
`ldr r1, [r1]` here to load the pointer from the vector table

carlescufi commented on arch/arm/core/irq_relay.S:
```diff
@@ -0,0 +1,112 @@
+GDATA(_main_stack)
+
+SECTION_FUNC(vector_relay_handler, __vector_relay_handler)
+	cpsid	i;
```
Do we really need to disable interrupts here? @cvinayak can you comment?

carlescufi replied:
remove this line, no need to disable interrupts

carlescufi commented on arch/arm/core/irq_relay.S:
```diff
@@ -0,0 +1,112 @@
+	 * be the real irq handle vector.
+	 * */
+
+	cpsie	i;
```
remove this line, no need to disable/re-enable interrupts

carlescufi commented:
@miyatsu I just tested this PR with the help of @nvlsianpu with the fixes I asked for by building mcuboot and samples/bluetooth/beacon and booting correctly with interrupt forwarding working fine.

carlescufi commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -182,4 +190,12 @@ config ZERO_LATENCY_IRQS
+config IRQ_SOFT_RELAY
+	bool
+	prompt "Enable Cortex-M0 Vector Table soft relay"
+	default y if CPU_CORTEX_M0 && !CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
```
Thinking about this, let's do the following for now:
```
default n
depends on CPU_CORTEX_M0 && !CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
```
I'm working on some approaches into how to enable this automatically for MCUboot, but for now let's not force it for everybody.

carlescufi commented:
Nice job @miyatsu. I have one last change I would like you to do and then we can merge.

miyatsu commented:
Rebase is done! @carlescufi

carlescufi commented:
@miyatsu thanks, let me test now and I will post results here.

carlescufi reviewed (approved):
This is ready for merging from my side, tested and verified. Thank you @miyatsu for your patience and work.

miyatsu commented:
@carlescufi Thanks for your generous help!
I will hand over the rest  enhancement work  to the community, thank you again.

mbolivar reviewed (approved):
approach seems reasonable to me

carlescufi commented:
@nashif this is ready to merge unless there are further objections

miyatsu committed:
arch: arm: Add SW_VECTOR_RELAY Kconfig flag

Eanble soft relay IRQ handler for Cortex-M0 target if this flag set
true.

Signed-off-by: Ding Tao <miyatsu@qq.com>

miyatsu committed:
arch: arm: vector_table: Support vector table soft relay

For SoCs that don't support vector table relocation in hardware, may not
support bootloader like mcuboot.

We introduce a way to relocate vector table in software by forwarding
the control of incoming IRQs to a new vector table which address is save
at fixed SRAM address.

User can change the data in that fixed SRAM address in order to relocate
vector table in software way.

Signed-off-by: Ding Tao <miyatsu@qq.com>

agross-oss reviewed (approved):
LGTM

carlescufi commented:
@d3zd3z I believe you reviewed an outdated version of the patches. `CONFIG_IS_BOOTLOADER` was never touched in the end.
